spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala

```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
-    * If raw record count is not available in checkpoints the method will calculate that count
-    * based on the provided raw dataframe.
+    * Adds metadata about the number of records in raw and source data by checking Atum's checkpoints first.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def addRawRecordCountToMetadata(df: DataFrame): Unit = {
+  private def addRawAndSourceRecordCountsToMetadata: Unit = {
     val checkpointRawRecordCount = getRawRecordCountFromCheckpoints
+    val checkpointSourceRecordCount = getSourceRecordCountFromCheckpoints
 
-    val rawRecordCount = checkpointRawRecordCount match {
-      case Some(num) => num
-      case None      => df.count
+    (checkpointRawRecordCount, checkpointSourceRecordCount) match {
+      case (Right(x), Right(y)) => {
+        Atum.setAdditionalInfo(s"raw_record_count" -> x.toString)
+        Atum.setAdditionalInfo(s"source_record_count" -> y.toString)
+      }
+      case (Right(_), Left(er)) => throw new ValidationException("Field errors", Seq(er))
+      case (Left(er), Right(_)) => throw new ValidationException("Field errors", Seq(er))
+      case (Left(er1), Left(er2)) => throw new ValidationException("Field errors", Seq(er1,er2))
     }
-    Atum.setAdditionalInfo(s"raw_record_count" -> rawRecordCount.toString)
   }
 
   /**
     * Gets the number of records in raw data by traversing Atum's checkpoints.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def getRawRecordCountFromCheckpoints: Option[Long] = {
+  private def getRawRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("row")
+  }
+
+  /**
+   * Gets the number of records in source data by traversing Atum's checkpoints.
+   *
+   * @return The number of records in a checkpoint corresponding to source data (if available)
+   */
+  private def getSourceRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("source")
+  }
+
+  private def getFieldRawRecordFromCheckpoints(field: String) = {
     import za.co.absa.atum.core.Constants._
-    val controlMeasure = Atum.getControlMeasure
-
-    val rawCheckpoint = controlMeasure
-      .checkpoints
-      .find(c => c.name.equalsIgnoreCase("raw") || c.workflowName.equalsIgnoreCase("raw"))
-
-    val measurement = rawCheckpoint.flatMap(chk => {
-      chk.controls.find(m => m.controlType.equalsIgnoreCase(controlTypeRecordCount))
-    })
-
-    measurement.flatMap(m =>
-      try {
-        val rawCount = m.controlValue.toString.toLong
-        // Basic sanity check
-        if (rawCount >= 0) {
-          Some(rawCount)
-        } else {
-          None
-        }
-      }
-      catch {
-        case NonFatal(_) => None
-      }
-    )
+    val optionResult = for {
+      checkpoint <- Atum.getControlMeasure
```

Comment:
I know this is not part of your code, but it can be made better.
If `getSourceRecordCountFromCheckpoints` gets a `ControlMeasure` instance, it can be unit tested for various cases. `Atum.getControlMeasure` can be called outsode the method.

Reply:
Makes sense